gcc/rust/typecheck/rust-hir-inherent-impl-check.h

```diff
@@ -0,0 +1,81 @@
+// Copyright (C) 2020-2024 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+#ifndef RUST_HIR_INHERENT_IMPL_ITEM_CHECK_H
+#define RUST_HIR_INHERENT_IMPL_ITEM_CHECK_H
+
+#include "rust-hir-item.h"
+#include "rust-hir-type-check-base.h"
+#include "rust-type-util.h"
+#include <vector>
+
+namespace Rust {
+namespace Resolver {
+
+class PrimitiveImplCheck : public TypeCheckBase
+{
+public:
+  static void go ()
+  {
+    PrimitiveImplCheck pass;
+
+    pass.scan ();
+  }
+
+private:
+  void scan ()
+
+  {
+    std::vector<HIR::ImplBlock *> possible_primitive_impl;
+    mappings.iterate_impl_blocks ([&] (HirId id, HIR::ImplBlock *impl) -> bool {
+      // filtering  trait-impl-blocks
+      if (impl->has_trait_ref ())
+	return true;
+      HirId impl_ty_id = impl->get_type ().get_mappings ().get_hirid ();
+      TyTy::BaseType *impl_type = nullptr;
+      if (!query_type (impl_ty_id, &impl_type))
+	return true;
+      if (is_primitive_type_kind (impl_type->get_kind ()))
+	{
+	  possible_primitive_impl.push_back (impl);
+	}
```

Comment:
I think this is a good start, but we must also check if the implementation is a lang item. In Rust 1.49, this is the implementation for the official `impl` of `i8` for example:

```rust

#[lang = "i8"]
impl i8 {
    int_impl! { i8, i8, u8, 8, -128, 127, "", "", 2, "-0x7e", "0xa", "0x12", "0x12", "0x48",
    "[0x12]", "[0x12]", "", "" }
}
```

because the standard library does define methods for the `i8` type so we still need to be able to do that somewhere.

Here is another primitive implementation if you want to take a look: https://github.com/rust-lang/rust/blob/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/ptr/mut_ptr.rs#L6

We can maybe address this in a later PR if this doesn't cause any regressions though

Reply:
:), i'm not sure if all these failed regressions just because missing lang item check, so many! sorry for that! :)  maybe  some of them need modify , I will take the time to do some reseach about your suggestions, thx!

Reply:
Yes some of the regressions are due to this, e.g. in gcc/testsuite/rust/compile/const-issue1440.rs where we define the primitive implementation for integers (which I assume will cause your error to trigger). but you'll need to fix the testcase as well to add the attribute